source/loader/ur_lib.cpp

```diff
@@ -42,20 +80,102 @@ __urdlllocal ur_result_t context_t::Init(ur_device_init_flags_t device_flags) {
         result = urInit();
     }
 
-    proxy_layer_context_t *layers[] = {
-        &ur_validation_layer::context,
-#if UR_ENABLE_TRACING
-        &ur_tracing_layer::context
-#endif
-    };
+    if (hLoaderConfig) {
+        enabledLayerNames.merge(hLoaderConfig->getEnabledLayerNames());
+    }
 
-    for (proxy_layer_context_t *l : layers) {
-        if (l->isEnabled()) {
-            l->init(&context->urDdiTable);
-        }
+    if (!enabledLayerNames.empty()) {
+        initLayers();
     }
 
     return result;
 }
 
+ur_result_t urLoaderConfigCreate(ur_loader_config_handle_t *phLoaderConfig) {
+    if (!phLoaderConfig) {
+        return UR_RESULT_ERROR_INVALID_NULL_POINTER;
+    }
+    *phLoaderConfig = new ur_loader_config_handle_t_;
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t urLoaderConfigRetain(ur_loader_config_handle_t hLoaderConfig) {
+    if (!hLoaderConfig) {
+        return UR_RESULT_ERROR_INVALID_NULL_HANDLE;
+    }
+    hLoaderConfig->incrementReferenceCount();
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t urLoaderConfigRelease(ur_loader_config_handle_t hLoaderConfig) {
+    if (!hLoaderConfig) {
+        return UR_RESULT_ERROR_INVALID_NULL_HANDLE;
+    }
+    if (hLoaderConfig->decrementReferenceCount() == 0) {
```

Comment:
This is a race, another thread might be concurrently calling release, causing double free on this object. You need to use `fetch_sub` here with `memory_order_acq_rel` (and `fetch_add` with the same flag when you increment).

Unless we don't care about thread safety here?

Reply:
totally didn't consider thread safety first time around, I've made a pass at making ref count atomic